library/src/main/java/com/bumptech/glide/load/resource/bitmap/GlideBitmapFactoryWrapper.java

```diff
@@ -0,0 +1,291 @@
+import java.io.InputStream;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Wrapper around {@link BitmapFactory} to work around known issues with {@link BitmapFactory}
+ * across Android SDK levels.
+ *
+ * <p>In particular, this class works around these known issues:
+ *
+ * <ul>
+ *   <li>Ultra HDR image single-channel gainmaps not being decoded on Android U when hardware
+ *       bitmaps are enabled. This issue is further described in
+ *       https://github.com/bumptech/glide/issues/5362.
+ * </ul>
+ *
+ * <p>New usages of {@link BitmapFactory} APIs within Glide should be added here rather than called
+ * directly.
+ */
+final class GlideBitmapFactoryWrapper {
+
+  private GlideBitmapFactoryWrapper() {}
+
+  /** Wrapper for {@link BitmapFactory#decodeStream}. */
+  @Nullable
+  public static Bitmap decodeStream(InputStream inputStream, BitmapFactory.Options options) {
+    if (VERSION.SDK_INT == VERSION_CODES.UPSIDE_DOWN_CAKE
+        && GainmapDecoderWorkaroundStateCalculator.needsGainmapDecodeWorkaround(options)) {
+      return safeDecodeHardwareBitmapWithGainmap(inputStream, options);
+    }
+    return BitmapFactory.decodeStream(inputStream, /* outPadding= */ null, options);
+  }
+
+  /** Wrapper for {@link BitmapFactory#decodeByteArray}. */
+  @Nullable
+  public static Bitmap decodeByteArray(byte[] bytes, BitmapFactory.Options options) {
+    if (VERSION.SDK_INT == VERSION_CODES.UPSIDE_DOWN_CAKE
+        && GainmapDecoderWorkaroundStateCalculator.needsGainmapDecodeWorkaround(options)) {
+      return safeDecodeHardwareBitmapWithGainmap(bytes, options);
+    }
+    return BitmapFactory.decodeByteArray(bytes, /* offset= */ 0, bytes.length, options);
+  }
+
+  /** Wrapper for {@link BitmapFactory#decodeFileDescriptor}. */
+  @Nullable
+  public static Bitmap decodeFileDescriptor(
+      FileDescriptor fileDescriptor, BitmapFactory.Options options) {
+    if (VERSION.SDK_INT == VERSION_CODES.UPSIDE_DOWN_CAKE
+        && GainmapDecoderWorkaroundStateCalculator.needsGainmapDecodeWorkaround(options)) {
+      return safeDecodeHardwareBitmapWithGainmap(fileDescriptor, options);
+    }
+    return BitmapFactory.decodeFileDescriptor(fileDescriptor, /* outPadding= */ null, options);
+  }
+
+  /**
+   * Returns a decoded bitmap for the input stream, ensuring that any associated gainmap is decoded
+   * without errors on Android U if it is a valid gainamp.
+   *
+   * <p>This method safely wraps BitmapFactory#decodeStream(InputStream, Rect, Options)} on Android
+   * U.
+   *
+   * @param inputStream for the bitmap to be decoded.
+   * @param options to be applied in the {@link BitmapFactory#decodeStream} call.
+   */
+  @RequiresApi(VERSION_CODES.UPSIDE_DOWN_CAKE)
+  @Nullable
+  private static Bitmap safeDecodeHardwareBitmapWithGainmap(
+      InputStream inputStream, Options options) {
+    Preconditions.checkArgument(options.inPreferredConfig == Config.HARDWARE);
+    options.inPreferredConfig = Bitmap.Config.ARGB_8888;
```

Comment:
you're setting options twice (once here and once inside the try). It should be the last statement before the `try {`

Applies to this and all the methods below

Reply:
Whoops thanks for catching this.